feat: add AtomOne v4 technical reference blog (#70)
atomone-hub/atom.one: content/english/blog/atomone-v4-technical-reference/index.md

}
```

The validation rules on `MsgUpdateParams` encode several constitutional constraints:

- The Oversight and Steering DAO addresses cannot be equal — to prevent a single entity from holding both roles simultaneously.
- Neither address can equal the module authority (the governance account) — to prevent governance itself from acting as a DAO.
- Neither DAO can have bonded or unbonding stake — enforcing the Constitution's prohibition on Core DAOs staking ATONE. This check queries the staking keeper at update time. For DAO addresses implemented as Interchain Accounts controlled from [gno.land](https://gno.land/), ICA accounts on AtomOne cannot stake, so this constraint is structurally guaranteed rather than merely point-in-time.

### 4.3 Steering DAO capabilities

**`MsgAnnotateProposal`:** Adds a free-text annotation to a proposal in the voting period. Annotations are stored on the `Proposal` object and visible in query responses. The Steering DAO may overwrite an existing annotation by setting `overwrite: true`. Annotation length is capped at 5,000 characters. The Steering DAO cannot annotate proposals that are not in the voting period.

> **`MsgEndorseProposal`:** Marks a proposal as endorsed, setting a boolean flag on the `Proposal` object. Endorsement is one-way — it cannot be revoked. Per the Constitution, a Steering DAO endorsement reduces the passing threshold for law proposals from Constitutional Majority (>90%) to Supermajority (>2/3). The mechanics of this threshold reduction are implemented in the SDK fork's tally logic, which reads the `Endorsed` field.

Reviewer: ```suggestion
**`MsgEndorseProposal`:** Marks a proposal as endorsed, setting a boolean flag on the `Proposal` object. Endorsement is one-way — it cannot be revoked. Per the Constitution, a Steering DAO endorsement reduces the passing threshold for law proposals from Constitutional Majority (>90%) to Supermajority (>2/3). The mechanics of this threshold reduction are implemented in the `x/gov`'s tally logic, which reads the `Endorsed` field.
```

### 5.4 Dynamic adjustment of $\eta$

The $\eta$ coefficient is adjusted weekly (via the `x/epochs` module) by the `AdjustNakamotoBonusCoefficient` function. The algorithm:

1. Retrieve all bonded validators and sort by voting power descending.
2. Divide into three equal groups of floor(N/3) validators each: high (top), medium (middle), low (bottom).
3. Compute the average bonded stake for the high and low groups.
4. If `high_avg / low_avg ≥ 3` (or if `low_avg == 0`): increase $\eta$ by `Step` (default 1%).
5. Otherwise: decrease $\eta$ by `Step`.
6. Clamp to `[MinimumCoefficient, MaximumCoefficient]` (defaults: 3% and 100%).

The 3× threshold reflects the following intuition: if the average stake of the top third of validators is at least three times the average of the bottom third, the network's stake distribution is sufficiently concentrated that the bonus should strengthen. Below that ratio, distribution is improving and the bonus can recede.

> The epoch-based triggering (via `x/epochs`) was a deliberate refactoring from the original fixed-block-interval implementation. Epochs decouple adjustment timing from block rate, making the behavior predictable from a calendar perspective and easier to reason about across chains with different block times. The distribution keeper implements `epochstypes.EpochHooks`, and `AfterEpochEnd` fires `AdjustNakamotoBonusCoefficient` whenever the epoch identifier matches `NakamotoBonus.PeriodEpochIdentifier`. The `x/epochs` module exists in the SDK fork and the distribution hooks are complete; the application-level wiring — registering the distribution keeper as an epoch hook consumer — is pending and will be completed before the v4 mainnet upgrade, analogously to the `x/dynamicfee` integration described in Section 6.

Reviewer: ```suggestion
The epoch-based triggering (via `x/epochs`) was a deliberate refactoring from the original fixed-block-interval implementation. Epochs decouple adjustment timing from block rate, making the behavior predictable from a calendar perspective and easier to reason about across chains with different block times. The distribution keeper implements `epochstypes.EpochHooks`, and `AfterEpochEnd` fires `AdjustNakamotoBonusCoefficient` whenever the epoch identifier matches `NakamotoBonus.PeriodEpochIdentifier`.
```

The conclusion is that Sybil attacks are only transiently profitable during the window before other delegators respond. Once the market equilibrates — which the bonus mechanism itself incentivizes — the attacker is left operating a more expensive (two nodes) configuration for no reward advantage. No additional protocol mechanism is required to prevent Sybil attacks; the economic dynamics of the bonus are self-correcting.

The `MaximumCoefficient` cap remains relevant as a governance guardrail: the analysis above assumes that 0.1% of total stake is mobile enough to respond. At very high $\eta$ values, the static gain during the pre-equilibration window grows large enough that it could be worth a short-term extraction. Keeping $\eta$ bounded by a governance-adjustable ceiling limits this window's upside without disabling the mechanism.

**Integer group division.** The group size is computed as `N / 3` (integer division). For N=100, the high and low groups each contain 33 validators; the middle 34 are unused in the comparison. For validator counts that are not multiples of 3, the groups are slightly unequal but the logic is otherwise unchanged.

> ## 6. Dynamic Fee Market: Migration to SDK Fork

Reviewer: ```suggestion
## 6. Dynamic Fee Market: Migration to AtomOne SDK
```

The `x/dynamicfee` module was introduced in v3. In v4, its canonical home moves from the AtomOne repository into the AtomOne SDK, where it is co-versioned with the other AtomOne-specific SDK modifications. This is purely a change of location and ownership, not of algorithm or interface: the module's behaviour is unchanged.

> The migration carries one substantive fix absent from the repository version: a guard against `ConsensusParams.Block.MaxGas` returning 0 or -1, the CometBFT representation of "unlimited block gas." Without this guard, the fee price calculation attempts to divide by a zero or negative target, producing incorrect behaviour. The fork introduces a `DefaultMaxBlockGas` parameter (default: 100,000,000 gas units), and the `GetMaxBlockGas` helper returns this value whenever the consensus parameter is absent or unlimited.

Reviewer: ```suggestion
The migration carries one substantive fix absent from the repository version: a guard against `ConsensusParams.Block.MaxGas` returning 0 or -1, the CometBFT representation of "unlimited block gas." Without this guard, the fee price calculation attempts to divide by a zero or negative target, producing incorrect behaviour. The updated version introduces a `DefaultMaxBlockGas` parameter (default: 100,000,000 gas units), and the `GetMaxBlockGas` helper returns this value whenever the consensus parameter is absent or unlimited.
```

### 7.5 Misbehaviour detection

The misbehaviour type carries two conflicting headers for the same height. `ValidateBasic` on `Misbehaviour` checks that:

- Both headers are individually valid (including their respective validator set hashes).
- Both headers reference the same chain ID.
- The commits in both headers are verified against the provided validator sets using `VerifyCommit`.
- `Header1.Height >= Header2.Height` (ordering convention).

If `CheckForMisbehaviour` confirms the two headers are conflicting valid signatures for the same height, `UpdateStateOnMisbehaviour` freezes the client at `FrozenHeight = (0, 1)`, preventing any further state updates. A frozen client requires a governance proposal (`RecoverClient`) to unfreeze.

The `ConsensusState` stored after each successful update contains:
- `Timestamp` — block time from the TM2 header
- `Root` — Merkle root derived from `AppHash` via ICS-23 `MerkleRoot`
- `NextValidatorsHash` — used in subsequent adjacent-header verification
- `LcType = "gno"` — identifies the light client type in the stored state

### 7.6 Scope and context

> The `10-gno` module is one component of a broader effort to establish full IBC v2 connectivity between AtomOne and Gno. The corresponding work on the Gno side — a Tendermint light client and IBC core logic implemented as Gno realms, along with a modular TypeScript relayer capable of operating across both chain types — is covered in detail in <PLACEHOLDER_LINK>.

Reviewer: ```suggestion
The `10-gno` module is one component of a broader effort to establish full IBC v2 connectivity between AtomOne and Gno. The corresponding work on the Gno side — a Tendermint light client and IBC core logic implemented as Gno realms, along with a modular TypeScript relayer capable of operating across both chain types — will be covered in detail in a dedicated article.
```

A summary of new message types introduced in v4 (all under `cosmos.gov.v1`):

```text
MsgCreateGovernor          { address, description }
MsgEditGovernor            { address, description }
MsgUpdateGovernorStatus    { address, status }       // ACTIVE or INACTIVE
MsgDelegateGovernor        { delegator_address, governor_address }
MsgUndelegateGovernor      { delegator_address }
```

And new query endpoints:

```text
QueryGovernor              { governor_address }
QueryGovernors             { status, pagination }
QueryGovernanceDelegation  { delegator_address }
QueryGovernanceDelegations { governor_address, pagination }
QueryGovernorValShares     { governor_address, pagination }
```

New `x/coredaos` messages:

```text
MsgAnnotateProposal        { annotator, proposal_id, annotation, overwrite }
MsgEndorseProposal         { endorser, proposal_id }
MsgExtendVotingPeriod      { extender, proposal_id }
MsgVetoProposal            { vetoer, proposal_id, burn_deposit }
MsgUpdateParams            { authority, params }
```

## 10. Security Audit

> The AtomOne v4 upgrade was subjected to a security audit carried out by [Oak Security](https://oaksecurity.io/). The full audit report is available at <PLACEHOLDER_LINK>.

Reviewer: ```suggestion
The AtomOne v4 upgrade was subjected to a security audit carried out by [Oak Security](https://oaksecurity.io/). The full audit report will be soon available at [github.com/atomone-hub/atomone/tree/main/docs](https://github.com/atomone-hub/atomone/tree/main/docs) as per previous versions audits.
```
